src/keyboard/mod.rs

```diff
@@ -569,59 +604,267 @@ where
-                    };
-                    let modifiers = state.mods_state.clone();
-                    user_impl.receive(
-                        Event::Key {
-                            serial,
-                            time,
-                            modifiers,
-                            rawkey: key,
-                            keysym: sym,
-                            state: key_state,
-                            utf8,
-                        },
-                        proxy,
-                    );
+                        if key_held == Some(key) {
+                            kill_chan.lock().unwrap().0.send(()).unwrap();
+                            key_held = None;
+                        }
+                        event_impl.receive(
+                            Event::Key {
+                                serial,
+                                time,
+                                modifiers,
+                                rawkey: key,
+                                keysym: sym,
+                                state: key_state,
+                                utf8: utf8.clone(),
+                            },
+                            proxy.clone(),
+                        );
+                    }
                 }
                 wl_keyboard::Event::Modifiers {
                     mods_depressed,
                     mods_latched,
                     mods_locked,
                     group,
                     ..
-                } => state.update_modifiers(mods_depressed, mods_latched, mods_locked, group),
+                } => {
+                    state.lock().unwrap().update_modifiers(mods_depressed, mods_latched, mods_locked, group);
+                    state_chan.lock().unwrap().0.send(()).unwrap();
+                },
                 wl_keyboard::Event::RepeatInfo { rate, delay } => {
-                    user_impl.receive(Event::RepeatInfo { rate, delay }, proxy);
+                    event_impl.receive(Event::RepeatInfo { rate, delay }, proxy);
+                    *system_repeat_timing.lock().unwrap() = (rate as u64, delay as u64);
                 }
             }
         },
     )
 }
+
+/// Implement a keyboard to automatically detect the keymap and send KeyRepeatEvents
+/// at set intervals
+///
+/// This requires you to provide an implementation to receive the events after they
+/// have been interpreted with the keymap. You must also provide an implementation to be called
+/// when KeyRepeatEvents are sent at intervals set by the KeyRepeatKind argument, this
+/// implementation can be called at anytime, independent of the dispatching of wayland events.
+///
+/// The keymap information will be loaded from the events sent by the compositor,
+/// as such you need to call this method as soon as you have created the keyboard
+/// to make sure this event does not get lost.
+///
+/// Returns an error if xkbcommon could not be initialized.
```

Comment:
Please add a note in this documentation and the one of `map_keyboard_rmlvo_with_repeat` that threads will be spawned in the background. I prefer this to be explicit.